Add document outlining our repo branching strategy

weshaggard opened the pull request:
The following document outlines the general branching strategy we plan to use across our azure-sdk-for-* repos. 

Tagging the engineering system team and the leads for the azure-sdk teams:
@Azure/azure-sdk-eng 
@kurtzeborn @AlexGhiondea @mayurid 
@salameer @sphibbs @lmazuel @yugangw-msft

Please review the document and provide feedback and add any additional reviews as needed.

bsiegel commented on docs/engineering-system/repo-branching.md:
```diff
@@ -0,0 +1,50 @@
+
+## Hotfix branches
+
+Under some circumstances we may need to service a specific version of the code with a hotfix and in these cases we should create a branch from the git tag that points at the specific version we want to hotfix with a name of `hotfix/<hotfix name/version>` and push it to the main repository. Once the branch is created the changes should be made and the version numbers for the package with the fix s ...
```
I'm not clear what "hotfix name" refers to here, do you have an example?

weshaggard replied:
I'll add a sentence about what a hotfix is. As for the versioning it is still just package-version but ultimately that version will map to whatever we version the hotfix.

AlexGhiondea commented on docs/engineering-system/repo-branching.md:
````diff
@@ -0,0 +1,50 @@
+Example of how to create a 1.0.0 release tag for the azure-keyvault package:
+
+```
+> git tag -a -m "1.0.0 release of azure-keyvault" azure-keyvault_1.0.0 HEAD
````
I know I can look up the git command, but I think having the arguments explained in this doc might help as well as the structure for the tag.

weshaggard replied:
I'm not sure it is that useful in this document as this is just an illustration but I will instead spell out the full names of the arguments to help a little.

AlexGhiondea commented on docs/engineering-system/repo-branching.md:
```diff
@@ -0,0 +1,50 @@
+
+In general there should not be a need for release branches because most releases will happen directly from master. In some cases there may be a need to lock down the branch to stabilize a package and in these cases we should create a release branch named `release/<release name>` and push it to the main repository. We do this to avoid ever locking down the master branch from taking other work. Aft ...
+
+There may be other circumstances where we need to release a preview of a package from another branch outside of master or a release branch. In those cases we should use a feature branch in the main repository (note that we should not do stable releases out of feature branches so ensure the version is correctly marked as a preview). The release should be done in the same way, including the creatio ...
```
Is there a way to enforce that we don't release non-preview bits from feature branches? I am thinking that in the build we can have some sort of validation that looks at the package version and the branch and ensures this rule.

weshaggard replied:
I don't think there is anything we can directly use here but we should consider potential options.

AlexGhiondea commented on docs/engineering-system/repo-branching.md:
```diff
@@ -0,0 +1,50 @@
+
+## Hotfix branches
+
+Under some circumstances we may need to service a specific version of the code with a hotfix and in these cases we should create a branch from the git tag that points at the specific version we want to hotfix with a name of `hotfix/<hotfix name/version>` and push it to the main repository. Once the branch is created the changes should be made and the version numbers for the package with the fix s ...
```
What happens if that hotfix has an issue and we need to 'hotfix the hotfix'?

weshaggard replied:
In that case you would still just bump to the next patch version. The only drawback will be that you cannot insert hotfixes in between hotfixes they will have to be a roll-up.

AlexGhiondea commented on docs/engineering-system/repo-branching.md:
```diff
@@ -0,0 +1,50 @@
+
+Under some circumstances we may need to service a specific version of the code with a hotfix and in these cases we should create a branch from the git tag that points at the specific version we want to hotfix with a name of `hotfix/<hotfix name/version>` and push it to the main repository. Once the branch is created the changes should be made and the version numbers for the package with the fix s ...
+
+Whenever we release a package that we need to support long term we need to ensure we have enough versioning space to release hotfixes at any point in the history of the package. To help ensure that we always need to bump at least the minor version (i.e. 2nd part) of the package because we will increment the patch (i.e. 3rd part) version each time we release a hotfix.
```
Is there any way to ensure this in the build system?

weshaggard replied:
There isn't any great way that I know of and I don't want to do anything to crazy. However if we can come up with something reasonable we should try and enforce it.

jsquire commented on docs/engineering-system/repo-branching.md:
```diff
@@ -0,0 +1,50 @@
+
+If there are a set of people that need to collaborate on the same set of changes before they can go into master then a feature branch can be pushed to the main repository for sharing. Collaborators can either work together to push changes to that branch or submit pull requests against it until it is ready to go to master. Once the feature work is done it should be rebased on top of master and the ...
+
+The feature branches pushed to the main repo should be named like `feature/<feature name>`
```
In this case is `feature` meant to be the literal string "feature?"  I'd offer the thought that the name of a particular SDK / service may be more helpful to locate a branch of interest when scanning through the list.   For example:

`servicebus/topic-retries`   versus  `feature/servicebus-topic-retries`.

weshaggard replied:
For branches that are persisted in the main repository I think prefixing them for the intention such as feature, release, or hotfix makes the intent clear for the type of work in that branch and also allows us to create policies around build triggers or releases from those branches, so I do like having the literal string feature.

I also like having the service in the name and nothing stops us from also having that convention as a sub branch name, something like `feature/servicebus/topic-retries`.

Do keep in mind I don't expect all branches for most work like standard bugs to follow this convention, for most standard bug fixes people will name the branch in their forks whatever they like. I'm mostly trying to ensure the main repository has a well structured set of branches that can be understood.

jsquire commented on docs/engineering-system/repo-branching.md:
```diff
@@ -0,0 +1,50 @@
+
+## Release branches
+
+There are potentially 3 different types of release branches in the order of preference:
```
**_nit:_**  List formatting doesn't seem to have gotten picked up in the GitHub rendering, for some reason _(unless the preview differs from what you're seeing)_.   Obviously, not important, but thought you may want to know.

see: [https://github.com/Azure/azure-sdk/blob/df5a9fb89788cf27ce18b3d08160aeadd38b479c/docs/engineering-system/repo-branching.md](https://github.com/Azure/azure-sdk/blob/df5a9fb89788cf27ce18b3d08160aeadd38b479c/docs/engineering-system/repo-branching.md)

weshaggard replied:
Thanks I missed that when I moved it to MD, I'll fix it.

jsquire commented on docs/engineering-system/repo-branching.md:
```diff
@@ -0,0 +1,50 @@
+
+## Feature branches
+
+Branches should be created off of master and in general should remain in your local or forked repository and when ready pull requests should be submitted to the main repository. Changes going into the master branch should generally be rebased on top of master.
```
**_nit:_** `and` in the first sentence of the paragraph seems awkward and may read better split into two sentences.   Something like "... should remain in your local (forked) repository.   When ready, pull requests..."

kurtzeborn replied:
Maybe "and should generally remain" instead of "and in general should remain"

jsquire commented on docs/engineering-system/repo-branching.md:
```diff
@@ -0,0 +1,50 @@
+The git branching and workflow strategy we will be using is mostly in line with [OneFlow](https://www.endoflineblog.com/oneflow-a-git-branching-model-and-workflow) with some slight variations called out below.
```
For what it's worth, I think going with OneFlow over GitFlow as a baseline is a good call.

weshaggard commented:
Tagging the larger team to get any other feedback folks might have @azure/azure-sdk-team please take a look.

johanste commented on docs/engineering-system/repo-branching.md:
```diff
@@ -0,0 +1,50 @@
+
+`master` is the main default branch that lives forever and should never be force pushed to. The master branch must always be in a working state where CI builds succeed (e.g. build, analyze, and tests passing).  All other branches are intentionally short lived and should be removed once they are no longer needed.
+
+Note that while master will always be in a buildable state it will not necessarily always represent the state of the latest published packages. To figure out the state of the code for a given released package you need to use the tag for that released package.
```
Under what circumstances would this not hold true?

bterlson replied:
I think this is probably talking about the public feed not internal feeds, so most master merges wouldn't have a corresponding release (changes build up until a release is made). IIUC that is :)

weshaggard replied:
That is correct I'm talking about what is published on the official package managers. While some packages like the management packages might always publish any changes merged into master the client packages most likely will not publish every change that goes into master.

bterlson commented on docs/engineering-system/repo-branching.md:
```diff
@@ -0,0 +1,50 @@
+
+## Feature branches
+
+For isolated work people should create branches off master and keep them in their local or forked repostory. The name for these branches is up to the individual working on the changes as these names will not show up anywhere (except perhaps a merge commit message) in the main repository. Once the work is ready the changes should be rebased on top of master and a pull request should be submitted t ...
```
Source branch for PRs shows up all over the GitHub UI at least... I wouldn't give the perception that these are in any sense "private" names.

weshaggard replied:
Fair I'll update to make it clear that they don't have to follow the `feature\<feature name>` branch naming convention.

weshaggard commented:
I believe I have addressed all the feedback that I've received and I've given folks enough time to provide feedback so I'm merging this PR now and we will design our engineering system around these principles. If folks have additional feedback please reach out to me or file an issue for me to address.
